core/src/main/scala/kafka/server/ControllerApis.scala

```diff
@@ -195,6 +198,61 @@ class ControllerApis(val requestChannel: RequestChannel,
       requestThrottleMs => createResponseCallback(requestThrottleMs))
   }
 
+  def handleDeleteTopics(request: RequestChannel.Request): Unit = {
+    if (!config.deleteTopicEnable) {
+      if (request.context.apiVersion() < 3) {
+        throw new InvalidRequestException("Topic deletion is disabled.")
+      } else {
+        throw new TopicDeletionDisabledException()
+      }
+    }
+    val deleteTopicsRequest = request.body[DeleteTopicsRequest]
+    val nameToId = new mutable.HashMap[String, Uuid]
+    deleteTopicsRequest.data().topicNames().iterator().asScala.foreach {
+      name => nameToId.put(name, Uuid.ZERO_UUID)
+    }
+    deleteTopicsRequest.data().topics().iterator().asScala.foreach {
+      nameAndId => nameToId.put(nameAndId.name(), nameAndId.topicId())
+    }
+    val (describable, deletable)  = {
+      if (authHelper.authorize(request.context, DELETE, CLUSTER, CLUSTER_NAME)) {
+        (nameToId.keySet, nameToId.keySet)
+      } else {
+        val authorizedDescribeTopics: Set[String] = authHelper.filterByAuthorized(
+          request.context, DESCRIBE, TOPIC, nameToId.keys)(n => n)
+        val authorizedDeleteTopics: Set[String] = authHelper.filterByAuthorized(
+          request.context, DELETE, TOPIC, nameToId.keys)(n => n)
+        (authorizedDescribeTopics, authorizedDeleteTopics)
+      }
+    }
+    def sendResponse(response: DeleteTopicsResponseData): Unit = {
+      nameToId.keysIterator.foreach {
```

Comment:
It's a bit odd to use iterators when calling `foreach` after. `iterator` is typically useful when you want to do `map` or other such operations lazily. `foreach` forces the iteration anyway.

Reply:
the main rationale for this is that I don't want scala to perform any copying, and I know for sure that if I give an iterator that it won't.

I wouldn't mind using Java's forEach directly but I couldn't find a good way to convert a scala closure to a java closure (maybe this is obvious and I missed it?)

Reply:
There is no copy when using foreach (both Java or Scala versions). And Scala lambdas are automatically converted to Java lambdas.